### ISSUE TITLE
Add async serialization support to XmlSerializer and DataContractSerializer

### ISSUE THREAD

**Issue by mconnew** [dotnet/runtime#32555]
To be able to support use of IAsyncEnumerable with WCF contracts (dotnet/wcf#4166), WCF needs support for async serialization. Without this, we would simply be moving which call stack a blocking thread is waiting on.

**Comment by xsoheilalizadeh** [comment]
Could we have this in .NET 5?

**Comment by Misiu** [comment]
I need both Serialize and Deserialize to be async.
Currently, we have `GetFromJsonAsync<T>` extension method for HttpClient.
I'd like to create a similar method, but for XML (yes, some old services are still returning XML)

**Comment by Misiu** [comment]
Any updates on this?

**Comment by vanillajonathan** [comment]
### Background and motivation
I would like to asynchronously deserialize streams.

### API Proposal
```diff
namespace System.Xml.Serialization;

public class XmlSerializer
{
+    public async Task<obect?> DeserializeAsync(Stream stream)
}
```

Or perhaps even with generics?
```diff
namespace System.Xml.Serialization;

public class XmlSerializer
{
+    public async Task<T?> DeserializeAsync<T>(Stream stream)
}
```

### API Usage
```csharp
var serializer = new XmlSerializer(typeof(Document));
var document = await serializer.DeserializeAsync(stream);
```

**Comment by quintonv** [comment]
I'm surprised this is still not in .Net 7. I have seen it requested quite a few times over the years...
I personally would love it too as I have previously encountered thread starvation in Web.Api in my Web Host

**Comment by plykkegaard** [comment]
Will it be available on NET 8?
Would be nice to have this
My current project handles a lot of xml deserialization

**Comment by elTRexx** [comment]
From what I read here, I am correct to suppose there is no XML asynchronous (de)serialization available ?

**Comment by davidfowl** [comment]
This is a monumental amount of work, it's not happening for .NET 8 (.NET 8 is done). I'm not sure it will happen for 9.

**Comment by plykkegaard** [comment]
No problem, thanks for replying and clarification

**Comment by mconnew** [comment]
Generally you don't need async deserialization. It's typically only needed when you have a list of items in the XML and you want to handle them one at a time. If you have a single object you wish to deserialize, you would asynchronously read from the input into a buffer and then synchronously deserialize from the buffer. If reading from a file, you know the size ahead of time so can allocate a buffer big enough. If someone over something like an Http request, hopefully you have a content length so you can allocate the correct sized buffer. WCF internally has a special Stream implementation for when HTTP chunked encoding is used where we have an array of byte array buffers to avoid over allocating or doing copy on grow. Then we use that Stream with the serializers. The scenario which is useful for WCF is when you are sending or returning `IEnumerable<Foo>` with your service. In this case your input is larger than the item you want to deserialize. You can't asynchronously read from the incoming data the right amount because there's no way to know you have enough data for the next object to be deserialized without implementing some XML parsing and doing look ahead. It's in this scenario where we end up doing sync over async with the serializer directly reading from the input stream. Unless you have a scenario like this, you don't need an async serializer. Chances are you can do something else like WCF does to avoid it.

**Comment by udlose** [comment]
@mconnew `JsonSerializer` has async Deserialization. 

Also, I wonder why `IXmlSerializable` only has sync methods for extending your own custom serialization.  I understand you can write your own interface, but it seems like this is basic enough where `IXmlSerializable` should include async signatures.

**Comment by mconnew** [comment]
There's a few problems with adding async signatures to IXmlSerializable, some resolvable, some not currently resolvable yet. The first is that you can't add any methods to an interface after it's been released. Take the situations where you have implemented IXmlSerializable as it is today in your app, then in a newer version of .NET new async methods are added. Someone runs your app against the newer .NET release and you'll get an exception from the CLR because your class which declares it implements IXmlSerializable is missing some methods. This can be resolved by creating a new interface for async variations.  
The next problem is that XmlSerializer and DataContractSerializer can't currently use an async implementation. They only have synchronous serialization implementations. If a new interface was created with async methods, the serializers would have to call them and block the thread while waiting for the task to complete. You don't gain any benefit from there being async methods.  

Once this issue has is implemented, we should look to add IAsyncXmlSerializble. It will need to be done thoughtfully though. Not all XmlReader/XmlWriter implementations support being called asynchronously. We would need to decide how to handle that. E.g. should we expect async and sync implementations of the read and write methods to exist and call and use the sync implementation when the reader/writer only support sync calls? Or should we leave that up to the implementor to decide? In your own app, you might know that only an async supporting reader/writer will be used so you don't want to be required to implement both. But that might still be okay as you can just throw NotImplementedException.

Basically it's a good idea once XmlSerializer/DataContractSerializer have an async implementation, but currently it's premature to add. Definitely something I hadn't explicitly thought of though so thanks for the suggestion.

**Comment by udlose** [comment]
@mconnew  thanks for your quick reply.  You'd mentioned 

> Generally you don't need async deserialization.

Why would you consider JsonSerialization different than XmlSerialization since JsonSerialization does have an async implementation?

**Comment by mconnew** [comment]
I explained most of it in the sentences that follow that statement, but I did miss out a bit of my rationale of why that approach is okay. Basically if you have an object graph to deserialize, the amount of memory that's needed for the buffer to hold the XML is going to be a similar order of magnitude as the final set of objects you will end up with, especially if you use a binary XML reader/writer with a persisted dictionary. If you have multiple items you will need to deserialize, eg receiving requests to a web service, using a buffer pool will prevent you stressing the GC. In the majority of cases, it's going to be fine to asynchronously read the entire XML into an in-memory buffer and then deserialize from that (which doesn't benefit from being async). As I mentioned, the only time that's not going to work is if you have a list of objects you are going to deserialize and you will handle them one at a time. For example, you have a time series result set of data from running an experiment and your data file is 500MB big. But you only need to process one result item at a time. In that case you would have a loop deserializing each one item, processing it, discarding the instance and going back for the next item. It's a lot harder to asynchronously read the input in that case and the naive implementation would end up doing synchronous reads. In many cases you can still work around that if it's a problem. If there's a known upper bound on how big each item is, you could create a custom stream which uses double buffering to ensure it always has enough data pre-read before deserializing the next item. In which case you have a loop which asynchronously ensure there's enough data pre-read, deserialize, process item, discard item, rinse and repeat.  

There's a small set of scenarios where you can't avoid synchronous reading, and they are going to be rare.

**Comment by mconnew** [comment]
To put it another way, having an async code path on a serializer allows you to write simpler code and not worry about things like item size when using a list, or buffer pooling. Not having it doesn't mean you have to read synchronously though.

**Comment by udlose** [comment]
@mconnew I guess my point was that async deserialization was added for json. Why should xml be any different? I may be streaming from a database for example in which case async is important.

**Comment by plykkegaard** [comment]
Xml has an exponential growth by each record compared to json
Eg UBL catalogs is one of the worst examples I have ever seen
I assume this has to be taken into consideration?

**Comment by mconnew** [comment]
@udlose, async deserialization wasn't added for json, the serializer you are referring to was written with an async implementation to begin with. It's a significantly different design. When XmlSerializer was first created, back in .NET Framework 1.1, it used codegen and called csc.exe to compile the serializer on demand. The async/await way of doing async didn't exist until 9 years later with .NET Framework 4.5. At some point a new implementation was added which used IL generation as it was faster to generate with a fallback to the old code gen implementation if that failed. When DataContractSerializer was created, it was made with only an IL generation mode. When everything was ported to .NET Core, there was no csc.exe available, the compiler is Roslyn, comes from nuget, and isn't available at runtime. So we had to drop the code gen implementation in XmlSerializer and use IL generation only. When .NET Native came around, there was a compile time code generator to generate the serializer to C#, but that had many limitations. As everything is being compiled as a single component all at once, name clashes between transitive types was a constant problem for developers which had no easy way to solve (we could likely solve it today using file scoped classes). It also ran into the problem of it being an imprecise mechanism to make sure all needed serializers were pre generated without also bloating the app by generating serializers for everything. The solution was a reflection based serializer. That reflects over the types and calls things like PropertyInfo.GetValue or FieldInfo.SetValue to do its work. I won't go into the details of what needs to change otherwise this would end up being a large blog post, but the design needs some changes to make it more performant. It's not bad, but it's worse than IL gen in most scenarios. So that's where we currently are today. As I'm sure you're aware, to do async/await, you need to do that all the way through the entire call stack. Have you ever looked at the state machine that the compiler generates when you make a method async and use an await statement? [This is a great blog post](https://devblogs.microsoft.com/dotnet/how-async-await-really-works/) with everything you ever wanted to know about who async/await works. Search for a code sample of an implementation of `MoveNext()`, it's not the first one, it's a few later, and you will see how complicated that is, when written in C#. Now imagine trying to rewrite a serializer which is implemented using Ref.Emit that emits individual IL instructions and now hand craft generating that state machine code dynamically for a serializer using IL, and trying to debug that implementation. It's not feasible without a large team and a large amount of time. The solution is going to be to fix the issues with the reflection based serializer to improve its perf (I believe we can get better than the IL implementation) and fix a few feature gaps we have, and then add async capability to that. As that is c# code instead of meta programming via IL generation, it's a lot more manageable scope of work. Also bear in mind that XML is a lot more complicated than JSON, so the implementation is larger than a JSON serializer would be too.   

Comparing adding async/await capability to a more complicated library that was designed to target a development/runtime environment significantly different from how things are today with a from scratch library designed for todays development environment is comparing apples to oranges.  

@plykkegaard, I can't think of a situation where Xml would grow exponentially compared to JSON. There are some scenarios where it would grow slower than JSON. For example, XML supports references, so if you have a lot of duplicated data, an XML document can be a lot smaller than the equivalent JSON would be, despite being a more verbose data format (which is largely negated if you use binary XML serialization). To achieve something similar with JSON you would have to manually express references in your .NET object model. Can you provide me a link to an example as it's possible my imagination just isn't good enough. I would expect a large complicated XML document to result in a large complicated .NET object model, but I'm willing to be proven wrong.

**Comment by mconnew** [comment]
@udlose, I don't see how streaming from a SQL server changes things. The approaches I mentioned earlier should be sufficient. Either you want to deserialize the entire Stream to a single object graph, in which case asynchronously download the entire Stream contents into memory and then deserialize the buffer. Use a buffer pool to remove GC load. If you have like a list of items, then pre-read an amount that is larger than an item can be and deserialize a single item, handle it, rinse and repeat.
